### PR TITLE
Add functional test for Kubernetes interop

### PR DESCRIPTION
**Pull request by rynowak**
# Description

This change adds a new function test for Kubernetes interop.

As part of this PR I added a new functional test run for the Kubernetes interop tests. It didn't feel like these belonged in our of our existing categories, and it's likely we'll add more in the future.

The major change in this PR is to handle synchronous operations correctly. The APIs we call are documented as asychronous, but can complete synchronously, which requires different handling in the code. I was able to address this without significantly complicating the logic.

There were a few other small fixes needed to get the tutorial code passing.

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #issue_number

### Summary
🧪🚀🛠️

This pull request adds Kubernetes support to Radius, by implementing Kubernetes controllers, custom resources, and functional tests. It also updates the documentation, the deployment chart, the CLI, and the error handling of the reconciler.

> _The pull request has many changes_
> _To test and improve Radius ranges_
> _It adds `kubernetes` tests_
> _And handles edge cases best_
> _With reconcilers and controllers exchanges_

### Walkthrough
*  Add Kubernetes functional tests and the environment template to verify the interoperation between Kubernetes manifests and Radius resources ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-f1b21adc52462ef91afa8e2bfa7927ab7c7ab2c25b6b29785aba25345bf9ccf8R1-R282), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-7ff04db92b609e2614cff7761401720302926bfb6b40f08908bd1ac3e9c8ae5dR1-R25))
*  Add the `kubernetes` name to the functional test matrix and the `test-functional-kubernetes` target to run the new tests ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-c79f364a9293abaaa8595776b74674e24bec6287834e63ab8aa7aec6a42f0dbcL299-R299), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-607e82dd7ce2736355a7703fb1a4a832dddb294c3683c4b51a5118218e379712L61-R61))
*  Add the `Connection` field to the `TestOptions` and `RPTestOptions` structs and change the type of the `Client` field to `client.WithWatch` to enable the functional tests to create API clients and watch custom resources ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-ca65669bbb1ffbb902e1e73c239e499e677a3efd9ab7e9f63ff826d14e604057R87), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-ca65669bbb1ffbb902e1e73c239e499e677a3efd9ab7e9f63ff826d14e604057R96-R98), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-415520be4a4b2230377a0f9c0625016d65206ef0768d9cfa76313be4387699ceL37-R37))
*  Import the `radapp.io/v1alpha3` API package and add the API group to the Kubernetes scheme in the `kubernetes` CLI package to enable the CLI to use the custom resources ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-bd86a3bd62b2c24bd445b9d7c4fbb5e48aa1aef0b3d137e992d80e129d6b5db1R37), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-bd86a3bd62b2c24bd445b9d7c4fbb5e48aa1aef0b3d137e992d80e129d6b5db1R53))
*  Change the return type of the `NewRuntimeClient` function and the call to `client.New` to support the `Watch` method in the `kubernetes` CLI package ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-bd86a3bd62b2c24bd445b9d7c4fbb5e48aa1aef0b3d137e992d80e129d6b5db1L85-R87), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-bd86a3bd62b2c24bd445b9d7c4fbb5e48aa1aef0b3d137e992d80e129d6b5db1L91-R93))
*  Rename the `controller` ClusterRole to `radius-controller` and add the `events` resource to avoid conflicts and enable event creation in the `deploy/Chart/templates/controller/rbac.yaml` file ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-0eb83f283dad161ea34486e10ead98d0c74fcaa93a13e8d182113ebb8de66c63L4-R4), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-0eb83f283dad161ea34486e10ead98d0c74fcaa93a13e8d182113ebb8de66c63R14))
*  Add the `controllers/` folder to the documentation of the code organization in the `docs/contributing/contributing-code/contributing-code-organization/README.md` file ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-5d21ae7061cfc67f297efd6dff13a428d728e4773925f3510f60c8e6d9e37783R31))
*  Handle the case where the delete, create, or update operation completes synchronously and update the annotations and status accordingly in the `reconcileDelete`, `reconcileUpdate`, `startPutOrDeleteOperationIfNeeded`, `startDeleteOperationIfNeeded`, `deleteResource`, `createOrUpdateResource`, `deleteContainer`, and `createOrUpdateContainer` functions in the `pkg/controller/reconciler` package ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L127-R129), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L301-R301), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L316-R316), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L358-R358), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L405-R414), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L451-R462), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L464-R483), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L553-R572), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-565f9bf43f08d107a3218cdeb8b17eab13363d0661a65535e80ec4ef114f09e5L384-R393), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-565f9bf43f08d107a3218cdeb8b17eab13363d0661a65535e80ec4ef114f09e5L402-R414), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-565f9bf43f08d107a3218cdeb8b17eab13363d0661a65535e80ec4ef114f09e5L415-R435), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-b2622e855cc3c8af44551c69c158d0400b6b2df3e3ebaefef18f7db96c6b66bbL166-R176), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-b2622e855cc3c8af44551c69c158d0400b6b2df3e3ebaefef18f7db96c6b66bbL188-R208), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-b2622e855cc3c8af44551c69c158d0400b6b2df3e3ebaefef18f7db96c6b66bbL217-R247), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-b2622e855cc3c8af44551c69c158d0400b6b2df3e3ebaefef18f7db96c6b66bbL239-R279))
*  Add suffixes to the error messages to distinguish between the errors from the `ResumeToken` and the `Result` methods of the pollers in the `reconcileDelete` and `reconcileUpdate` functions in the `pkg/controller/reconciler` package ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L301-R301), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L316-R316), [link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L358-R358))
*  Ignore the error if the secret to be deleted does not exist in the `cleanupDeployment` function in the `pkg/controller/reconciler` package ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L553-R572))
*  Handle the case where the annotations are nil in the `SetupWithManager` function in the `pkg/controller/reconciler` package ([link](https://github.com/radius-project/radius/pull/6488/files?diff=unified&w=0#diff-87a7dfa06c174a6b41b671b2cfffb84c81e481881baec866a026e7dd00db8195L617-R636))

### CODE REVIEW

File: .github/workflows/functional-test.yaml [hunk 1]
```diff
@@ -296,7 +296,7 @@ jobs:
       fail-fast: true
       matrix:
         os: [ubuntu-latest]
-        name: [ucp,shared,msgrp,daprrp,samples]
+        name: [ucp,kubernetes,shared,msgrp,daprrp,samples]
```
Comment (rynowak) [inline .github/workflows/functional-test.yaml:299]: Adding a new run to the matrix, this doesn't really belong in the other categories.

File: deploy/Chart/templates/controller/rbac.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
-  name: controller
+  name: radius-controller
```
Comment (rynowak) [inline deploy/Chart/templates/controller/rbac.yaml:4]: This is a cluster-scoped object. Renamed to avoid conflict with other packages.

File: deploy/Chart/templates/controller/rbac.yaml [hunk 3]
```diff
@@ -11,6 +11,7 @@ rules:
   resources:
   - namespaces
   - secrets
+  - events
```
Comment (rynowak) [inline deploy/Chart/templates/controller/rbac.yaml:14]: Needed for diagnostics. This is the best way to enable troubleshooting for users.

File: pkg/cli/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -49,6 +50,7 @@ func init() {
 	_ = apiextv1.AddToScheme(Scheme)
 	_ = clientgoscheme.AddToScheme(Scheme)
 	_ = contourv1.AddToScheme(Scheme)
+	_ = radappiov1alpha3.AddToScheme(Scheme)
```
Comment (rynowak) [inline pkg/cli/kubernetes/kubernetes.go:53]: Added so we can interact with the recipe CRD type in our code. There's no real drawback to this.

File: pkg/cli/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -82,13 +84,13 @@ func NewClientset(context string) (*k8s.Clientset, *rest.Config, error) {
...
 // NewRuntimeClient creates a kubernetes client using a given context and scheme.
-func NewRuntimeClient(context string, scheme *k8s_runtime.Scheme) (client.Client, error) {
+func NewRuntimeClient(context string, scheme *k8s_runtime.Scheme) (client.WithWatch, error) {
 	merged, err := NewCLIClientConfig(context)
 	if err != nil {
 		return nil, err
 	}
 
-	c, err := client.New(merged, client.Options{Scheme: scheme})
+	c, err := client.NewWithWatch(merged, client.Options{Scheme: scheme})
```
Comment (rynowak) [inline pkg/cli/kubernetes/kubernetes.go:93]: Changed to make testing easier. We need to watch in tests.

File: pkg/controller/reconciler/deployment_reconciler.go [hunk 6]
```diff
@@ -124,9 +124,9 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		}
 	}
 
-	// If the Deployment is being deleted **or** if Radius is not enabled, then we should
+	// If the Deployment is being deleted **or** if Radius is no longer enabled, then we should
 	// clean up any Radius state.
-	if deployment.DeletionTimestamp != nil || annotations.Configuration == nil {
+	if deployment.DeletionTimestamp != nil || (annotations.Configuration == nil && annotations.Status != nil) {
```
Comment (rynowak) [inline pkg/controller/reconciler/deployment_reconciler.go:129]: Logic bug was causing this code to process all deployments, not just those where Radius was being disabled.

File: pkg/controller/reconciler/deployment_reconciler.go [hunk 7]
```diff
@@ -298,7 +298,7 @@ func (r *DeploymentReconciler) reconcileUpdate(ctx context.Context, deployment *
 		// We've successfully started an operation. Update the status and requeue.
 		token, err := updatePoller.ResumeToken()
 		if err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to get operation token: %w", err)
+			return ctrl.Result{}, fmt.Errorf("failed to get operation token1: %w", err)
```
Comment (rynowak) [inline pkg/controller/reconciler/deployment_reconciler.go:301]: Will revert this, was for my debugging.

File: pkg/controller/reconciler/deployment_reconciler.go [hunk 8]
```diff
@@ -402,10 +402,16 @@ func (r *DeploymentReconciler) startPutOrDeleteOperationIfNeeded(ctx context.Con
...
 			return nil, nil, false, err
+		} else if poller != nil {
+			return nil, poller, false, nil
 		}
 
-		return nil, poller, false, err
-	} else if !annotations.IsUpToDate() {
+		// Deletion completed synchronously.
+		annotations.Status.Container = ""
+	}
```
Comment (rynowak) [inline pkg/controller/reconciler/deployment_reconciler.go:411]: All of the code that handles pollers is changing like this. We need to handle the case where the operation completed synchronously and successfully.

File: pkg/controller/reconciler/deployment_reconciler.go [hunk 9]
```diff
@@ -550,7 +569,7 @@ func (r *DeploymentReconciler) cleanupDeployment(ctx context.Context, deployment
 
 	secretName := client.ObjectKey{Namespace: deployment.Namespace, Name: fmt.Sprintf("%s-connections", deployment.Name)}
 	err := r.Client.Delete(ctx, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Namespace: secretName.Namespace, Name: secretName.Name}})
-	if err != nil {
+	if err != nil && !apierrors.IsNotFound(err) {
```
Comment (rynowak) [inline pkg/controller/reconciler/deployment_reconciler.go:572]: Missing a check for this edge case.

File: pkg/controller/reconciler/util.go [hunk 10]
```diff
@@ -163,7 +163,17 @@ func deleteResource(ctx context.Context, radius RadiusClient, resourceID string)
...
-	return poller, nil
+	if !poller.Done() {
+		return poller, nil
+	}
+
+	// Handle synchronous completion
+	_, err = poller.Result(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (rynowak) [inline pkg/controller/reconciler/util.go:174]: Per conversation with SDK experts this is how we detect synchronous completion.

**Review by lakshmimsft — APPROVED** [review]
